### PR TITLE
[SSDK-487] Improve AddressAutofill, PlaceAutocomplete efficiency by skipping prefetching

### PR DESCRIPTION
**Pull request by aokj4ck**
### Description

Fixes [SSDK-487](https://mapbox.atlassian.net/browse/SSDK-487)

- Update AddressAutofill.resolve(suggestions:) to _only_ perform `autofill/v1/suggest` requests and to abstain from fetching the `retrieve` endpoint
- Update PlaceAutocomplete.select(suggestions:) to _only_ perform `search/v1/suggest` requests and to abstain from fetching the `retrieve` endpoint
- Requesting only `suggest` removes the suggestion coordinates and estimated times. Perform a call to `retrieve` in order to fetch those values.
- Disable SwiftLint rules for the time being
- Add helper function to convert CoreAddress to AddressAutofill.AddressComponents
- Change AddressAutofill.Suggestion.coordinate to optional
- Change PlaceAutocomplete.Suggestion.coordinate and PlaceAutocomplete.Result.coordinate to optional
- Add Suggestion.Underlying field and enum to capture original Core API response for future API requests or validation

### Checklist

- [x] Update `CHANGELOG`

### Screenshots

| Before | After |
| -- | -- |
| Simulator Screen Shot - iPhone 14 Pro - 2024-01-09 at 12 31 30 | Simulator Screenshot - iPhone 15 - 2023-12-20 at 13 57 19 |
| Simulator Screenshot - iPhone 15 - 2023-12-20 at 13 55 20 | Simulator Screen Shot - iPhone 14 Pro - 2024-01-09 at 12 15 32 |

[SSDK-487]: https://mapbox.atlassian.net/browse/SSDK-487?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: Sources/MapboxSearch/PublicAPI/Use Cases/Place Autocomplete/PlaceAutocomplete.swift [hunk 1]
```diff
@@ -291,52 +291,28 @@ private extension PlaceAutocomplete {
...
-        }
-        
-        dispatchGroup.notify(queue: .main) {
-            let results = resolvedSuggestions.compactMap({ $0 })
-            if results.isEmpty {
-                completion(.failure(resolvingError ?? SearchError.responseProcessingFailed))
-            } else {
-                completion(.success(results))
-            }
+            return Suggestion(name: matchingName,
```
Comment (kried) [inline Sources/MapboxSearch/PublicAPI/Use Cases/Place Autocomplete/PlaceAutocomplete.swift:304]: From what I understand,`name` and `matchingName` are different properties.
```
/** Names matched to the \a languages array. */
@property (nonatomic, readonly, nonnull, copy) NSArray<NSString *> *names;
/** The feature name, as matched by the search algorithm. */
@property (nonatomic, readonly, nullable, copy) NSString *matchingName;
```

Before this pull request the values `searchSuggestion.names.first ?? ""` was used, now `matchingName` is used. Is this change expected?

File: Sources/MapboxSearch/PublicAPI/Use Cases/Address Autofill/AddressAutofill.swift [hunk 2]
```diff
@@ -197,56 +200,59 @@ private extension AddressAutofill {
...
     ) {
-        guard !suggestions.isEmpty else {
-            return completion(.success([]))
+        let resolvedSuggestions = suggestions.compactMap { result -> Suggestion? in
+            guard let address = result.addresses?.first,
+                  let resultAddress = try? address.toAutofillComponents() else {
+                return nil
+            }
+
+            let matchingName = result.matchingName ?? " "
```
Comment (kried) [inline Sources/MapboxSearch/PublicAPI/Use Cases/Address Autofill/AddressAutofill.swift:209]: Is `matchingName` expected to be used as `name`? Is ` " "` valid default value?

File: Sources/MapboxSearch/PublicAPI/Use Cases/Address Autofill/AddressAutofill.swift [hunk 3]
```diff
@@ -105,13 +105,16 @@ public extension AddressAutofill {
...
         userActivityReporter.reportActivity(forComponent: "address-autofill-suggestion-select")
 
-        let result = AddressAutofill.Result(
-            name: suggestion.name,
-            formattedAddress: suggestion.formattedAddress,
-            coordinate: suggestion.coordinate,
-            addressComponents: suggestion.addressComponents
-        )
-        completion(.success(result))
+        guard case let .suggestion(coreSearch, coreOptions) = suggestion.underlying else {
```
Comment (kried) [inline Sources/MapboxSearch/PublicAPI/Use Cases/Address Autofill/AddressAutofill.swift:108]: What about the case when `underlying` is `.result` case?  Are such cases invalid? Should then `.result` enum case be removed?

File: Sources/MapboxSearch/PublicAPI/Use Cases/Address Autofill/Models/Suggestion/AddressAutofill+Suggestion.swift [hunk 4]
```diff
@@ -11,21 +11,35 @@ public extension AddressAutofill {
         /// Textual representation of the address.
         public let formattedAddress: String
         
-        /// Address geographic point.
-        public let coordinate: CLLocationCoordinate2D
+        /// Address geographic point. May be nil.
+        public let coordinate: CLLocationCoordinate2D?
```
[the remark this answers is not in the document]
Comment (aokj4ck) [reply Sources/MapboxSearch/PublicAPI/Use Cases/Address Autofill/Models/Suggestion/AddressAutofill+Suggestion.swift:15]: Yes this is necessary to fulfill the purpose of the ticket. When fetching from the `retrieve` endpoint directly the coordinate would be non-nil but from the `suggest` endpoint it must be optional.

File: Sources/MapboxSearch/PublicAPI/Use Cases/Place Autocomplete/PlaceAutocomplete.swift (hunk 1 above)
Comment (aokj4ck) [reply Sources/MapboxSearch/PublicAPI/Use Cases/Place Autocomplete/PlaceAutocomplete.swift:304]: Oh that's probably better to use `name`

File: Sources/MapboxSearch/PublicAPI/Use Cases/Address Autofill/AddressAutofill.swift (hunk 3 above)
Comment (aokj4ck) [reply Sources/MapboxSearch/PublicAPI/Use Cases/Address Autofill/AddressAutofill.swift:108]: - For AddressAutofill the `Underlying.result` enum case is still used by the `fetchSuggestions(using: completion:)` function which calls the `reverseGeocoding` API. 
- Similarly for PlaceAutocomplete the result is still necessary to use the `select(suggestion: completion:)` function.

File: CHANGELOG.md [hunk 5]
```diff
@@ -6,6 +6,9 @@ Guide: https://keepachangelog.com/en/1.0.0/
 
 ## Unreleased
 
+- [Address Autofill]: removed pre-fetching for suggestions
```
[the remark this answers is not in the document]
Comment (aokj4ck) [reply CHANGELOG.md:9]: 👍

File: Sources/Demo.xcodeproj/project.xcworkspace/xcshareddata/swiftpm/Package.resolved
[the remark this answers is not in the document]
Comment (aokj4ck) [reply Sources/Demo.xcodeproj/project.xcworkspace/xcshareddata/swiftpm/Package.resolved:1]: Hm interesting. I haven't made any changes around the `brand` property and this is resolved to MapboxCommon 23.8.6. You can test this out by pinning the Package.swift maximum version to `let commonMaxVersion = Version("23.6.1")` and the builds still succeed for each target.

[the remark this answers is not in the document]
Comment (aokj4ck) [reply Sources/Demo.xcodeproj/project.xcworkspace/xcshareddata/swiftpm/Package.resolved:1]: Or could this be caused/affected by the `mapboxSearchSDKVersion = "1.0.0-rc.8"` change?

File: Sources/MapboxSearch/PublicAPI/Use Cases/Address Autofill/AddressAutofill.swift (hunk 2 above)
Comment (aokj4ck) [reply Sources/MapboxSearch/PublicAPI/Use Cases/Address Autofill/AddressAutofill.swift:209]: I'll change this to `guard let name = result.names.first` so we always know that a valid name is available for non-nil results.

File: Sources/MapboxSearch/PublicAPI/Use Cases/Address Autofill/AddressAutofill.swift (hunk 3 above)
Comment (aokj4ck) [reply Sources/MapboxSearch/PublicAPI/Use Cases/Address Autofill/AddressAutofill.swift:108]: I have an idea for some more improvements in this function later today or tomorrow

File: Sources/Demo/PlaceAutocompleteDetailsViewController.swift [hunk 6]
```diff
@@ -80,18 +80,20 @@ private extension PlaceAutocompleteResultViewController {
...
+        annotation.coordinate = coordinate
         annotation.title = result.name
 
         mapView.addAnnotation(annotation)
     }
     
     func showSuggestionRegion() {
-        guard result != nil else { return }
-        
+        guard result != nil, let coordinate = result.coordinate else { return }
```
Comment (azarovalex) [inline Sources/Demo/PlaceAutocompleteDetailsViewController.swift:93]: small nit
```suggestion
        guard let coordinate = result.coordinate else { return }
```

File: Sources/MapboxSearch/PublicAPI/Use Cases/Address Autofill/Models/Suggestion/CoreAddress+AddressComponents.swift [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+
+        houseNumber.map { components.append(.init(kind: .houseNumber, value: $0)) }
+        street.map { components.append(.init(kind: .street, value: $0)) }
+        neighborhood.map { components.append(.init(kind: .neighborhood, value: $0)) }
+        locality.map { components.append(.init(kind: .locality, value: $0)) }
+        postcode.map { components.append(.init(kind: .postcode, value: $0)) }
+        place.map { components.append(.init(kind: .place, value: $0)) }
+        district.map { components.append(.init(kind: .district, value: $0)) }
+        region.map { components.append(.init(kind: .region, value: $0)) }
+        country.map { components.append(.init(kind: .country, value: $0)) }
```
Comment (azarovalex) [inline Sources/MapboxSearch/PublicAPI/Use Cases/Address Autofill/Models/Suggestion/CoreAddress+AddressComponents.swift:21]: Can we rewrite this as a for loop, using `CaseIterable`?
Right now there is a chance we can forget to update this function after adding a new case to AddressComponent.Kind

File: Tests/MapboxSearchUITests/Integration/CategorySuggestionsIntegrationTestCase.swift [hunk 8]
```diff
@@ -34,8 +34,8 @@ class CategorySuggestionsIntegrationTestCase: MockServerTestCase {
         
         let suggestions = app.categorySuggestionsTableView
         waitForHittable(suggestions, message: "CategorySuggestionsController tableView not hittable")
-        XCTAssertTrue(suggestions.cells.count > 0, "No Category suggestions results")
-        
+        XCTAssertGreaterThan(suggestions.cells.count, 0, "Category suggestions results should not be empty")
```
Comment (azarovalex) [inline Tests/MapboxSearchUITests/Integration/CategorySuggestionsIntegrationTestCase.swift:37]: Tbh, XCTAssertTrue used to be more readable😄

File: Tests/MapboxSearchUITests/VisibilityTestCase.swift [hunk 9]
```diff
@@ -70,7 +70,7 @@ class VisibilityTestCase: BaseTestCase {
         searchBar.swipeUp()
         
         let categoriesTableView = waitForHittable(app.tables["CategoriesTableViewSource.tableView"])
-        XCTAssertTrue(categoriesTableView.cells.count > 0, "Categories TableView empty")
+        XCTAssertGreaterThan(categoriesTableView.cells.count, 0, "Categories TableView should not be empty")
```
Comment (azarovalex) [inline Tests/MapboxSearchUITests/VisibilityTestCase.swift:73]:
```suggestion
        XCTAssertFalse(categoriesTableView.cells.isEmpty, "Categories TableView should not be empty")
```

**Review by azarovalex — APPROVED** [review]

File: Sources/MapboxSearch/PublicAPI/Use Cases/Address Autofill/Models/Suggestion/CoreAddress+AddressComponents.swift (hunk 7 above)
Comment (aokj4ck) [reply Sources/MapboxSearch/PublicAPI/Use Cases/Address Autofill/Models/Suggestion/CoreAddress+AddressComponents.swift:21]: We would still have to line up each `components` array member with the field from `CoreAddress` to assign the data to the right enum case. This is built [following the pattern from AddressAutofill.Suggestion+SearchResult](https://github.com/mapbox/mapbox-search-ios/blob/main/Sources/MapboxSearch/PublicAPI/Use%20Cases/Address%20Autofill/Models/Suggestion/AddressAutofill.Suggestion%2BSearchResult.swift#L41).

**Review by globaltrouble — APPROVED** [review]

**Pull request merged**